File: lua/vibe-coding/tests/fixtures/fixture_loader.lua
-- Test fixture loader for VibeApplyPatch tests
-- Provides utilities to load and manage test fixtures

local M = {}

-- Get the directory where this file is located
local fixtures_dir = vim.fn.fnamemodify(debug.getinfo(1).source:sub(2), ':p:h')

--- Trim leading and trailing whitespace from a string
-- This handles the case where fixture content is defined with [[ ]] brackets
-- and we want to trim the leading/trailing newlines and whitespace for cleaner editing
-- @param str string: The string to trim
-- @return string: The trimmed string
local function trim_content(str)
  if type(str) ~= 'string' then
    return str
  end

  -- Only trim the specific newlines that come after [[ and before ]]
  -- First, trim a single trailing newline at the start (after [[)
  str = str:gsub('^%s*\n', '')
  -- Then, trim a single leading newline at the end (before ]])
  str = str:gsub('\n%s*$', '')

  return str
end

--- Load content from an external text file
-- @param fixture_dir string: Directory containing the fixture
-- @param filename string: Name of the text file to load
-- @return string|nil: Content of the file, or nil if file doesn't exist
local function load_external_content(fixture_dir, filename)
  local content_path = fixture_dir .. '/' .. filename

  if vim.fn.filereadable(content_path) ~= 1 then
    return nil
  end

  local content = table.concat(vim.fn.readfile(content_path), '\n')
  return content
end

--- Try to load external content using convention-based naming
-- @param fixture_dir string: Directory containing the fixture
-- @param fixture_name string: Base name of the fixture (without .lua extension)
-- @param content_type string: Type of content ('original', 'diff', 'expected')
-- @return string|nil: Content of the file, or nil if file doesn't exist
local function try_load_conventional_content(fixture_dir, fixture_name, content_type)
  local filename = fixture_name .. '_' .. content_type .. '.txt'
  return load_external_content(fixture_dir, filename)
end

--- Load a single fixture file
-- @param fixture_path string: Path relative to fixtures directory (e.g., 'simple/basic_modification')
-- @return table: The fixture data
function M.load_fixture(fixture_path)
  -- Remove .lua extension if provided
  local clean_path = fixture_path:gsub('%.lua$', '')

  local full_path = fixtures_dir .. '/' .. clean_path .. '.lua'

  if vim.fn.filereadable(full_path) ~= 1 then
    error('Fixture file not found: ' .. full_path)
  end

  local ok, fixture = pcall(dofile, full_path)
  if not ok then
    error('Failed to load fixture: ' .. fixture_path .. '\nError: ' .. fixture)
  end

  -- Get the directory containing this fixture file and fixture name
  local fixture_dir = vim.fn.fnamemodify(full_path, ':h')
  local fixture_name = vim.fn.fnamemodify(full_path, ':t:r')

  -- Load content from external files if specified (takes priority over inline content)
  if fixture.original_content_file then
    local content = load_external_content(fixture_dir, fixture.original_content_file)
    if not content then
      error('External original content file not found: ' .. fixture_dir .. '/' .. fixture.original_content_file)
    end
    fixture.original_content = content
  elseif not fixture.original_content then
    -- Try convention-based loading: {fixture_name}_original.txt
    fixture.original_content = try_load_conventional_content(fixture_dir, fixture_name, 'original')
  end

  if fixture.diff_content_file then
    local content = load_external_content(fixture_dir, fixture.diff_content_file)
    if not content then
      error('External diff content file not found: ' .. fixture_dir .. '/' .. fixture.diff_content_file)
    end
    fixture.diff_content = content
  elseif not fixture.diff_content then
    -- Try convention-based loading: {fixture_name}_diff.txt
    fixture.diff_content = try_load_conventional_content(fixture_dir, fixture_name, 'diff')
  end

  if fixture.expected_content_file then
    local content = load_external_content(fixture_dir, fixture.expected_content_file)
    if not content then
      error('External expected content file not found: ' .. fixture_dir .. '/' .. fixture.expected_content_file)
    end
    fixture.expected_content = content
  elseif not fixture.expected_content and fixture.should_succeed then
    -- Try convention-based loading: {fixture_name}_expected.txt
    fixture.expected_content = try_load_conventional_content(fixture_dir, fixture_name, 'expected')
  end

  -- Trim whitespace from content fields for cleaner editing (only if loaded inline)
  if fixture.original_content and not fixture.original_content_file then
    fixture.original_content = trim_content(fixture.original_content)
  end
  if fixture.diff_content and not fixture.diff_content_file then
    fixture.diff_content = trim_content(fixture.diff_content)
  end
  if fixture.expected_content and not fixture.expected_content_file then
    fixture.expected_content = trim_content(fixture.expected_content)
  end

  -- Validate fixture structure
  M.validate_fixture(fixture, fixture_path)

  -- Add metadata
  fixture._path = fixture_path
  fixture._full_path = full_path

  return fixture
end

--- Load all fixtures from a category directory
-- @param category string: Category name (e.g., 'simple', 'complex', 'error_cases')
-- @return table: Array of fixture data
function M.load_category(category)
  local category_dir = fixtures_dir .. '/' .. category

  if vim.fn.isdirectory(category_dir) ~= 1 then
    error('Category directory not found: ' .. category_dir)
  end

  local fixtures = {}
  local files = vim.fn.glob(category_dir .. '/*.lua', false, true)

  for _, file in ipairs(files) do
    local fixture_name = vim.fn.fnamemodify(file, ':t:r')
    local fixture_path = category .. '/' .. fixture_name

    local ok, fixture = pcall(M.load_fixture, fixture_path)
    if ok then
      table.insert(fixtures, fixture)
    else
      error('Failed to load fixture: ' .. fixture_path .. '\nError: ' .. fixture)
    end
  end

  -- Sort by name for consistent ordering
  table.sort(fixtures, function(a, b)
    return a.name < b.name
  end)

  return fixtures
end

--- Load all fixtures from all categories
-- @return table: Array of all fixture data organized by category
function M.load_all()
  local categories = { 'simple', 'complex', 'error_cases' }
  local all_fixtures = {}

  for _, category in ipairs(categories) do
    local category_fixtures = M.load_category(category)
    for _, fixture in ipairs(category_fixtures) do
      fixture._category = category
      table.insert(all_fixtures, fixture)
    end
  end

  return all_fixtures
end

--- Get all available categories
-- @return table: Array of category names
function M.get_categories()
  local categories = {}
  local dirs = vim.fn.glob(fixtures_dir .. '/*', false, true)

  for _, dir in ipairs(dirs) do
    if vim.fn.isdirectory(dir) == 1 then
      local category_name = vim.fn.fnamemodify(dir, ':t')
      -- Skip non-category directories
      if category_name ~= 'README.md' and not category_name:match '%.lua$' then
        table.insert(categories, category_name)
      end
    end
  end

  table.sort(categories)
  return categories
end

--- Validate fixture structure
-- @param fixture table: The fixture data to validate
-- @param fixture_path string: Path for error reporting
function M.validate_fixture(fixture, fixture_path)
  fixture_path = fixture_path or 'unknown'

  -- Required fields
  local required_fields = { 'name', 'description', 'should_succeed' }

  for _, field in ipairs(required_fields) do
    if fixture[field] == nil then
      error('Missing required field "' .. field .. '" in fixture: ' .. fixture_path)
    end
  end

  -- Content fields - must have content after loading (inline, explicit file, or convention-based)
  if not fixture.original_content then
    error('Missing required "original_content" in fixture: ' .. fixture_path .. ' (not found inline, via explicit file, or convention-based)')
  end

  if not fixture.diff_content then
    error('Missing required "diff_content" in fixture: ' .. fixture_path .. ' (not found inline, via explicit file, or convention-based)')
  end

  -- Type validation
  if type(fixture.name) ~= 'string' then
    error('Field "name" must be a string in fixture: ' .. fixture_path)
  end

  if type(fixture.description) ~= 'string' then
    error('Field "description" must be a string in fixture: ' .. fixture_path)
  end

  if type(fixture.should_succeed) ~= 'boolean' then
    error('Field "should_succeed" must be a boolean in fixture: ' .. fixture_path)
  end

  if fixture.original_content and type(fixture.original_content) ~= 'string' then
    error('Field "original_content" must be a string in fixture: ' .. fixture_path)
  end

  if fixture.diff_content and type(fixture.diff_content) ~= 'string' then
    error('Field "diff_content" must be a string in fixture: ' .. fixture_path)
  end

  -- Conditional validation
  if fixture.should_succeed then
    if not fixture.expected_content then
      error('Missing required "expected_content" when should_succeed=true in fixture: ' .. fixture_path .. ' (not found inline, via explicit file, or convention-based)')
    end
    if fixture.expected_content and type(fixture.expected_content) ~= 'string' then
      error('Field "expected_content" must be a string in fixture: ' .. fixture_path)
    end
  else
    if not fixture.expected_error_pattern then
      error('Field "expected_error_pattern" is required when should_succeed=false in fixture: ' .. fixture_path)
    end
    if type(fixture.expected_error_pattern) ~= 'string' then
      error('Field "expected_error_pattern" must be a string in fixture: ' .. fixture_path)
    end
  end

  -- Optional fields validation
  if fixture.tags and type(fixture.tags) ~= 'table' then
    error('Field "tags" must be a table in fixture: ' .. fixture_path)
  end

  if fixture.file_path and type(fixture.file_path) ~= 'string' then
    error('Field "file_path" must be a string in fixture: ' .. fixture_path)
  end
end

-- Helper function to setup manual mocking
local function setup_file_mocks(read_content, write_result)
  local vibe = require 'vibe-coding'
  local read_calls = {}
  local write_calls = {}
  
  local original_read = vibe.Utils.read_file_content
  local original_write = vibe.Utils.write_file
  
  vibe.Utils.read_file_content = function(filepath)
    table.insert(read_calls, filepath)
    return read_content, nil
  end
  
  vibe.Utils.write_file = function(filepath, content)
    table.insert(write_calls, {filepath = filepath, content = content})
    return write_result or true, nil
  end
  
  return {
    read_calls = read_calls,
    write_calls = write_calls,
    restore = function()
      vibe.Utils.read_file_content = original_read
      vibe.Utils.write_file = original_write
    end
  }
end

--- Create a test case from a fixture
-- @param fixture table: The fixture data
-- @return function: A test function that can be used with testing frameworks
function M.create_test_case(fixture)
  return function()
    local vibe = require 'vibe-coding'

    local mocks = setup_file_mocks(fixture.original_content)

    -- Parse and apply the diff
    local parsed_diff, parse_err = vibe.VibePatcher.parse_diff(fixture.diff_content)

    if fixture.should_succeed then
      -- Test should succeed
      assert.is_nil(parse_err, 'Failed to parse diff: ' .. (parse_err or ''))
      assert.is_not_nil(parsed_diff, 'Parsed diff should not be nil')

      local success, msg = vibe.VibePatcher.apply_diff(parsed_diff)

      assert.is_true(success, 'Expected apply_diff to succeed: ' .. (msg or ''))

      -- Handle special case for file deletion (message is in expected_content)
      if fixture.expected_content:match '^Skipped file deletion' then
        assert.are.equal(fixture.expected_content, msg)
        assert.are.equal(0, #mocks.write_calls, 'Expected write_file to not be called')
      else
        assert.is_true(msg:find 'Successfully applied' ~= nil, 'Message should contain "Successfully applied"')
        assert.is_true(msg:find 'hunks to' ~= nil, 'Message should contain "hunks to"')

        -- Verify file was written with expected content
        assert.are.equal(1, #mocks.write_calls, 'Expected write_file to be called once')
        local expected_lines = vim.split(fixture.expected_content, '\n', { plain = true, trimempty = false })
        assert.are.equal(fixture.file_path or parsed_diff.new_path, mocks.write_calls[1].filepath)
        assert.are.same(expected_lines, mocks.write_calls[1].content)
      end
    else
      -- Test should fail
      if parsed_diff then
        local success, msg = vibe.VibePatcher.apply_diff(parsed_diff)
        assert.is_false(success, 'Expected apply_diff to fail')
        assert.is_true(
          (msg or ''):find(fixture.expected_error_pattern) ~= nil,
          'Error message should contain expected pattern'
        )
        assert.are.equal(0, #mocks.write_calls, 'Expected write_file to not be called')
      else
        -- Parse error is also acceptable for error cases
        assert.is_true(
          (parse_err or ''):find(fixture.expected_error_pattern) ~= nil,
          'Parse error should contain expected pattern'
        )
      end
    end
    
    mocks.restore()
  end
end

return M

File: lua/vibe-coding/python_impl/fixture_loader.py
#!/usr/bin/env python3
"""
Python fixture loader that can read Lua-based test fixtures.
Provides cross-language compatibility with the existing Lua fixture system.
"""

import re
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple


class FixtureLoader:
    """Loads test fixtures from the Lua-based fixture system."""

    def __init__(self, fixtures_dir: Optional[str] = None):
        """Initialize the fixture loader.

        Args:
            fixtures_dir: Path to fixtures directory. If None, auto-detects based on this file's location.
        """
        if fixtures_dir is None:
            # Auto-detect fixtures directory relative to this file
            current_dir = Path(__file__).parent
            self.fixtures_dir = current_dir.parent / "tests" / "fixtures"
        else:
            self.fixtures_dir = Path(fixtures_dir)

        if not self.fixtures_dir.exists():
            raise FileNotFoundError(
                f"Fixtures directory not found: {self.fixtures_dir}"
            )

    def _parse_lua_fixture(self, lua_content: str) -> Dict[str, Any]:
        """Parse a Lua fixture file and extract the fixture data.

        Args:
            lua_content: Content of the Lua fixture file

        Returns:
            Dictionary containing the fixture data
        """
        # Remove 'return ' prefix and trailing commas/semicolons
        lua_content = lua_content.strip()
        if lua_content.startswith("return "):
            lua_content = lua_content[7:]

        # Parse the Lua table structure
        fixture = {}

        # Extract simple string/boolean/number fields
        patterns = {
            "name": r"name\s*=\s*['\"]([^'\"]*)['\"]",
            "description": r"description\s*=\s*['\"]([^'\"]*)['\"]",
            "should_succeed": r"should_succeed\s*=\s*(true|false)",
            "file_path": r"file_path\s*=\s*['\"]([^'\"]*)['\"]",
            "expected_error_pattern": r"expected_error_pattern\s*=\s*['\"]([^'\"]*)['\"]",
            "original_content_file": r"original_content_file\s*=\s*['\"]([^'\"]*)['\"]",
            "diff_content_file": r"diff_content_file\s*=\s*['\"]([^'\"]*)['\"]",
            "expected_content_file": r"expected_content_file\s*=\s*['\"]([^'\"]*)['\"]",
            "original_content": r"original_content\s*=\s*['\"]([^'\"]*)['\"]",  # Include single-line original_content
        }

        # Also check for single-line expected_content assignments
        expected_content_single = re.search(
            r"expected_content\s*=\s*['\"]([^'\"]*)['\"]", lua_content
        )
        if expected_content_single:
            fixture["expected_content"] = expected_content_single.group(1)

        for field, pattern in patterns.items():
            match = re.search(pattern, lua_content)
            if match:
                value = match.group(1)
                if field == "should_succeed":
                    fixture[field] = value == "true"
                else:
                    fixture[field] = value

        # Extract tags array
        tags_match = re.search(r"tags\s*=\s*\{([^}]*)\}", lua_content)
        if tags_match:
            tags_content = tags_match.group(1)
            tags = re.findall(r"['\"]([^'\"]*)['\"]", tags_content)
            fixture["tags"] = tags

        # Extract multi-line string content (using [[ ]] syntax)
        content_patterns = {
            "original_content": r"original_content\s*=\s*\[\[(.*?)\]\]",
            "diff_content": r"diff_content\s*=\s*\[\[(.*?)\]\]",
            "expected_content": r"expected_content\s*=\s*\[\[(.*?)\]\]",
        }

        for field, pattern in content_patterns.items():
            match = re.search(pattern, lua_content, re.DOTALL)
            if match:
                content = match.group(1)
                # Trim leading/trailing newlines (similar to Lua trim_content function)
                content = re.sub(r"^\s*\n", "", content)
                content = re.sub(r"\n\s*$", "", content)
                fixture[field] = content

        return fixture

    def _load_external_content(self, fixture_dir: Path, filename: str) -> Optional[str]:
        """Load content from an external text file.

        Args:
            fixture_dir: Directory containing the fixture
            filename: Name of the text file to load

        Returns:
            Content of the file, or None if file doesn't exist
        """
        content_path = fixture_dir / filename
        if not content_path.exists():
            return None

        with open(content_path, "r", encoding="utf-8") as f:
            return f.read()

    def _try_load_conventional_content(
        self, fixture_dir: Path, fixture_name: str, content_type: str
    ) -> Optional[str]:
        """Try to load external content using convention-based naming.

        Args:
            fixture_dir: Directory containing the fixture
            fixture_name: Base name of the fixture (without .lua extension)
            content_type: Type of content ('original', 'diff', 'expected')

        Returns:
            Content of the file, or None if file doesn't exist
        """
        filename = f"{fixture_name}_{content_type}.txt"
        return self._load_external_content(fixture_dir, filename)

    def _validate_fixture(self, fixture: Dict[str, Any], fixture_path: str) -> None:
        """Validate fixture structure.

        Args:
            fixture: The fixture data to validate
            fixture_path: Path for error reporting

        Raises:
            ValueError: If fixture is invalid
        """
        # Required fields
        required_fields = ["name", "description", "should_succeed"]

        for field in required_fields:
            if field not in fixture:
                raise ValueError(
                    f'Missing required field "{field}" in fixture: {fixture_path}'
                )

        # Content fields - must have content after loading (but can be empty string)
        if "original_content" not in fixture:
            raise ValueError(
                f'Missing required "original_content" in fixture: {fixture_path}'
            )

        if "diff_content" not in fixture:
            raise ValueError(
                f'Missing required "diff_content" in fixture: {fixture_path}'
            )

        # Type validation
        if not isinstance(fixture["name"], str):
            raise ValueError(
                f'Field "name" must be a string in fixture: {fixture_path}'
            )

        if not isinstance(fixture["description"], str):
            raise ValueError(
                f'Field "description" must be a string in fixture: {fixture_path}'
            )

        if not isinstance(fixture["should_succeed"], bool):
            raise ValueError(
                f'Field "should_succeed" must be a boolean in fixture: {fixture_path}'
            )

        # Conditional validation
        if fixture["should_succeed"]:
            if "expected_content" not in fixture:
                raise ValueError(
                    f'Missing required "expected_content" when should_succeed=True in fixture: {fixture_path}'
                )
        else:
            if "expected_error_pattern" not in fixture:
                raise ValueError(
                    f'Field "expected_error_pattern" is required when should_succeed=False in fixture: {fixture_path}'
                )

    def load_fixture(self, fixture_path: str) -> Dict[str, Any]:
        """Load a single fixture file.

        Args:
            fixture_path: Path relative to fixtures directory (e.g., 'simple/basic_modification')

        Returns:
            The fixture data dictionary
        """
        # Remove .lua extension if provided
        clean_path = fixture_path
        if clean_path.endswith(".lua"):
            clean_path = clean_path[:-4]

        full_path = self.fixtures_dir / f"{clean_path}.lua"

        if not full_path.exists():
            raise FileNotFoundError(f"Fixture file not found: {full_path}")

        # Load and parse the Lua fixture file
        with open(full_path, "r", encoding="utf-8") as f:
            lua_content = f.read()

        fixture = self._parse_lua_fixture(lua_content)

        # Get the directory containing this fixture file and fixture name
        fixture_dir = full_path.parent
        fixture_name = full_path.stem

        # Load content from external files if specified (takes priority over inline content)
        for content_type in ["original", "diff", "expected"]:
            content_file_key = f"{content_type}_content_file"
            content_key = f"{content_type}_content"

            if content_file_key in fixture:
                # Explicit external file specified
                content = self._load_external_content(
                    fixture_dir, fixture[content_file_key]
                )
                if content is None:
                    raise FileNotFoundError(
                        f"External {content_type} content file not found: {fixture_dir}/{fixture[content_file_key]}"
                    )
                fixture[content_key] = content
            elif content_key not in fixture:
                # Try convention-based loading: {fixture_name}_{content_type}.txt
                content = self._try_load_conventional_content(
                    fixture_dir, fixture_name, content_type
                )
                if content is not None:
                    fixture[content_key] = content

        # Validate fixture structure
        self._validate_fixture(fixture, fixture_path)

        # Add metadata
        fixture["_path"] = fixture_path
        fixture["_full_path"] = str(full_path)

        return fixture

    def load_category(self, category: str) -> List[Dict[str, Any]]:
        """Load all fixtures from a category directory.

        Args:
            category: Category name (e.g., 'simple', 'complex', 'error_cases')

        Returns:
            List of fixture data dictionaries
        """
        category_dir = self.fixtures_dir / category

        if not category_dir.exists():
            raise FileNotFoundError(f"Category directory not found: {category_dir}")

        fixtures = []
        lua_files = list(category_dir.glob("*.lua"))

        for lua_file in lua_files:
            fixture_name = lua_file.stem
            fixture_path = f"{category}/{fixture_name}"

            try:
                fixture = self.load_fixture(fixture_path)
                fixtures.append(fixture)
            except Exception as e:
                raise RuntimeError(
                    f"Failed to load fixture: {fixture_path}\nError: {e}"
                )

        # Sort by name for consistent ordering
        fixtures.sort(key=lambda x: x["name"])

        return fixtures

    def load_all(self) -> List[Dict[str, Any]]:
        """Load all fixtures from all categories.

        Returns:
            List of all fixture data dictionaries organized by category
        """
        categories = ["simple", "complex", "error_cases", "integration"]
        all_fixtures = []

        for category in categories:
            try:
                category_fixtures = self.load_category(category)
                for fixture in category_fixtures:
                    fixture["_category"] = category
                    all_fixtures.append(fixture)
            except FileNotFoundError:
                # Category might not exist, skip it
                continue

        return all_fixtures

    def get_categories(self) -> List[str]:
        """Get all available categories.

        Returns:
            List of category names
        """
        categories = []

        for item in self.fixtures_dir.iterdir():
            if item.is_dir() and not item.name.startswith("."):
                categories.append(item.name)

        categories.sort()
        return categories


def create_test_data_from_fixture(
    fixture: Dict[str, Any],
) -> Tuple[str, str, str, bool, Optional[str]]:
    """Convert fixture data to format used by Python tests.

    Args:
        fixture: Fixture data dictionary

    Returns:
        Tuple of (original_content, diff_content, expected_content, should_succeed, expected_error_pattern)
    """
    return (
        fixture["original_content"],
        fixture["diff_content"],
        fixture.get("expected_content", ""),
        fixture["should_succeed"],
        fixture.get("expected_error_pattern"),
    )


# Convenience functions for easy usage
def load_fixture(
    fixture_path: str, fixtures_dir: Optional[str] = None
) -> Dict[str, Any]:
    """Load a single fixture file.

    Args:
        fixture_path: Path relative to fixtures directory
        fixtures_dir: Optional fixtures directory path

    Returns:
        Fixture data dictionary
    """
    loader = FixtureLoader(fixtures_dir)
    return loader.load_fixture(fixture_path)


def load_category(
    category: str, fixtures_dir: Optional[str] = None
) -> List[Dict[str, Any]]:
    """Load all fixtures from a category.

    Args:
        category: Category name
        fixtures_dir: Optional fixtures directory path

    Returns:
        List of fixture data dictionaries
    """
    loader = FixtureLoader(fixtures_dir)
    return loader.load_category(category)


def load_all_fixtures(fixtures_dir: Optional[str] = None) -> List[Dict[str, Any]]:
    """Load all fixtures from all categories.

    Args:
        fixtures_dir: Optional fixtures directory path

    Returns:
        List of all fixture data dictionaries
    """
    loader = FixtureLoader(fixtures_dir)
    return loader.load_all()
